### ISSUE TITLE
Video Reconstruction based on WAN

### ISSUE THREAD

**Issue by ArlenCHEN** [DSL-Lab/UniEdit-Flow#2]
Hi @caramel-jiao, 

Thanks for your great work! Can you also provide a script to perform video reconstruction based on WAN?

Much appreciated!

**Comment by caramel-jiao** [comment]
Of course! Here is the code below.
You can provide your description of the video through `--prompt "<DESCRIPTION>"` to conduct conditional inversion and reconstruction, or set `--prompt ""` for unconditional situation.
Additionally, you can also try to modify line-60 from `prompt=args.prompt,` to `prompt="",` to have a different mixed setting.

To be honest, the reconstruction results are perhaps not so satisfactory, probably because videos are complex data. Based on my experience, under the conditional setting, if you can provide as detailed and accurate descriptions as possible, you will get much better reconstruction results.

```python
import os 
import argparse

import torch
from diffusers.utils import export_to_video, load_video
from diffusers import AutoencoderKLWan, WanPipeline
from diffusers.schedulers import FlowMatchEulerDiscreteScheduler

from schedulers.UniInvEulerScheduler import UniInvEulerScheduler
from utils import seed_everything, encode_video, decode_video


if __name__ == '__main__':
    parser = argparse.ArgumentParser("Video inversion and reconstruction based on diffusers")
    parser.add_argument("--video_path", type=str)
    parser.add_argument("--save_folder", type=str)
    parser.add_argument("--prompt", type=str, default="")
    
    parser.add_argument("--flow_shift", type=float, default=3.0)
    
    parser.add_argument("--step", type=int, default=50, help='1~1000')
    parser.add_argument("--seed", default=8, type=int)
    parser.add_argument("--fps", type=int, default=16)
    parser.add_argument("--model_path", type=str, default="Wan-AI/Wan2.1-T2V-1.3B-Diffusers")
    
    args = parser.parse_args()
    
    # preparation
    os.makedirs(args.save_folder, exist_ok=True)
    seed_everything(args.seed)
    device = 'cuda'
    
    # load VAE and pipeline separately to avoid OOM
    vae = AutoencoderKLWan.from_pretrained(args.model_path, subfolder="vae", torch_dtype=torch.float32)
    pipe = WanPipeline.from_pretrained(args.model_path, vae=vae, torch_dtype=torch.bfloat16)
    pipe.to('cuda')
    
    # load scheduler
    invert_scheduler = UniInvEulerScheduler(num_train_timesteps=1000, shift=args.flow_shift)
    scheduler = FlowMatchEulerDiscreteScheduler(num_train_timesteps=1000, shift=args.flow_shift)
    
    # load video
    video = load_video(args.video_path)
    height = video[0].size[1]
    width = video[0].size[0]
    num_frames = len(video)
    
    # AE
    video_latents = encode_video(pipe, vae, video)
    ae_video = decode_video(pipe, vae, video_latents)
    export_to_video(
        ae_video, 
        os.path.join(args.save_folder, 'ae_video.mp4'), 
        fps=args.fps
    )
    
    # invert
    pipe.scheduler = invert_scheduler
    invert_noise_latents = pipe(
        prompt=args.prompt,
        height=height,
        width=width,
        num_frames=num_frames,
        num_inference_steps=args.step,
        guidance_scale=1.0,
        output_type='latent',
        latents=video_latents,
    ).frames
    invert_noise_video = decode_video(pipe, vae, invert_noise_latents)
    export_to_video(
        invert_noise_video, 
        os.path.join(args.save_folder, 'invert_noise_video.mp4'), 
        fps=args.fps
    )
    
    # recon
    pipe.scheduler = scheduler
    recon_latents = pipe(
        prompt=args.prompt,
        height=height,
        width=width,
        num_frames=num_frames,
        num_inference_steps=args.step,
        guidance_scale=1.0,
        output_type='latent',
        latents=invert_noise_latents,
        ).frames
    recon_video = decode_video(pipe, vae, recon_latents)
    export_to_video(
        recon_video, 
        os.path.join(args.save_folder, 'result_recon_video.mp4'), 
        fps=args.fps
    )
```

**Comment by ArlenCHEN** [comment]
thank you so much!

**Comment by ArlenCHEN** [comment]
hi @caramel-jiao,

I noticed one thing -- the result is messed up if I try to run a generation from random noise -- by just removing the arguments of the pipe `latents=invert_noise_latents`. Below are the used prompt and the resulting video. This is weird to me, because it is just equivalent to the original wanx model. Do you have any idea about this?

Prompt: an intense volleyball match between two teams, one wearing green and the other in white, taking place in a well-lit indoor gymnasium. The players are actively engaged in the game, with one team attempting to serve the ball while the other team prepares to defend. The crowd in the background is cheering and watching the match with excitement. The scene is filled with dynamic movements, including jumping, diving, and quick reactions from the players, showcasing the competitive nature of the sport. The main subjects are the volleyball players, divided into two teams, one wearing green and the other in white. The players are positioned around the court, with some actively engaged in the game, such as serving, diving, and jumping to hit the ball. The players are wearing athletic uniforms and are positioned strategically around the net and the court. The referee, dressed in a white shirt and black pants, stands near the net, closely observing the game. The camera is stationary, providing a wide-angle view of the entire court and the surrounding area, capturing the entire scene without any noticeable movement or zooming.

Resulting video:
https://github.com/user-attachments/assets/058b1599-b94a-4aae-a85f-4625574d813b

**Comment by caramel-jiao** [comment]
I've just had the same try as you and I guess perhaps the `guidance_scale` is the key. In [https://github.com/Wan-Video/Wan2.1/blob/main/generate.py](https://github.com/Wan-Video/Wan2.1/blob/main/generate.py), their official generation script adopt a guidance_scale of 5.0. You can adjust this parameter and I think you will get your wanted results.

The reason why we utilize 1.0 as guidance_scale is for accurate reconstruction (i.e., without introducing different sampling in the reconstruction process). And in terms of editing, Uni-Edit does not need CFG thus reducing computional complex and better background preservation. So in our scripts you may only find guidance_scale=1.0 instead of some values facilitates generation.

BTW, you can also try some different flow_shift values, I think it will also result in various performance.

**Comment by caramel-jiao** [comment]
Moreover, you can try some high-order schedulers. [Wan_Diffusers](https://huggingface.co/Wan-AI/Wan2.1-T2V-1.3B-Diffusers/blob/main/scheduler/scheduler_config.json) utilizes UniPC (unified prediction-correction scheduler) with 2-order (I think it's similar to Heun maybe), which requires more NFEs but obtains better generation results.

**Comment by ArlenCHEN** [comment]
yes, you are right. turning guidance_scale to 5.0 gives a much better result.

will post more interesting and useful findings in this thread. thanks a lot!

expect to see your new work if any is coming soon.

**Comment by ArlenCHEN** [comment]
@caramel-jiao 
I tested some videos for reconstruction. Some are good, some are bad.
Good:
- Prompt: an intense volleyball match between two teams, one wearing green and the other in white, taking place in a well-lit indoor gymnasium. The players are actively engaged in the game, with one team attempting to serve the ball while the other team prepares to defend. The crowd in the background is cheering and watching the match with excitement. The scene is filled with dynamic movements, including jumping, diving, and quick reactions from the players, showcasing the competitive nature of the sport. The main subjects are the volleyball players, divided into two teams, one wearing green and the other in white. The players are positioned around the court, with some actively engaged in the game, such as serving, diving, and jumping to hit the ball. The players are wearing athletic uniforms and are positioned strategically around the net and the court. The referee, dressed in a white shirt and black pants, stands near the net, closely observing the game. The camera is stationary, providing a wide-angle view of the entire court and the surrounding area, capturing the entire scene without any noticeable movement or zooming.
- Video: https://github.com/user-attachments/assets/e32eb4f1-6a72-4889-a1ee-95e80450bdee

Bad:
- Prompt: a moment of celebration on a soccer field where two players from the same team are seen embracing and cheering after a successful play. The players are wearing blue and white uniforms with the number 12 and 13, respectively. They are surrounded by other teammates who are also celebrating, with some running towards the camera and others standing in the background. The scene is filled with joy and camaraderie, showcasing the spirit of teamwork and sportsmanship. The main subjects are two soccer players in blue and white uniforms, with the number 12 and 13 on their backs. They are in the center of the frame, hugging each other and smiling. The player with the number 12 is wearing blue shorts and socks, while the player with the number 13 is wearing white shorts and socks. They are surrounded by other players in similar uniforms, who are also celebrating. The players are moving towards the camera, with some running and others standing still. The background consists of a soccer field with green grass and white boundary lines. There are some advertisement banners visible on the sides of the field. The scene is set during the daytime, with sunlight casting shadows on the field. The weather appears to be clear and sunny, contributing to the vibrant atmosphere of the celebration. The camera is stationary, capturing the scene from a wide-angle view, allowing the players and their movements to be fully visible within the frame.
- Video: https://github.com/user-attachments/assets/ceeacb22-22fc-47b0-b4c6-9ee4efd99672

Are those results normal for reconstruction?

**Comment by ArlenCHEN** [comment]
The input videos for the above recon are:
- Good: https://github.com/user-attachments/assets/f86b7f87-4067-4d0c-8d4b-0852d4db44b7
- Bad: https://github.com/user-attachments/assets/3c82c7ee-5993-4ebc-a8f4-9c94d7a1469e

**Comment by caramel-jiao** [comment]
I think they're normal. On the one hand, it seems that Wan 1.3B is not good at handling complex human interactions, and on the other hand, I used to gain some even worse results😂. So I think your prompts should be already sufficient enough and the results should have been correct for the provided implementation.

**Comment by ArlenCHEN** [comment]
Have you ever tried 14b for reconstruction?

**Comment by caramel-jiao** [comment]
So far, I haven't tried it. Maybe it would provide better capability for inversion and reconstruction I think.
